# Exercise 02 -  OLAP Cubes - Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://video.udacity-data.com/topher/2021/August/61120e06_pagila-3nf/pagila-3nf.png)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](https://video.udacity-data.com/topher/2021/August/61120d38_pagila-star/pagila-star.png)

Start by creating and connecting to the database by running the cells below.

In [6]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

createdb: database creation failed: ERROR:  database "pagila_star" already exists
 set_config 
------------
 
(1 row)

psql:Data/pagila-star.sql:41: ERROR:  type "mpaa_rating" already exists
psql:Data/pagila-star.sql:51: ERROR:  type "year" already exists
psql:Data/pagila-star.sql:68: ERROR:  function "_group_concat" already exists with same argument types
psql:Data/pagila-star.sql:85: ERROR:  function "film_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:102: ERROR:  function "film_not_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:147: ERROR:  function "get_customer_balance" already exists with same argument types
psql:Data/pagila-star.sql:169: ERROR:  function "inventory_held_by_customer" already exists with same argument types
psql:Data/pagila-star.sql:206: ERROR:  function "inventory_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:224: ERROR:  function "last_day" already exists with same argument 

### Connect to the local database where Pagila is loaded

In [7]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Start with a simple cube
TODO: Write a query that calculates the revenue (sales_amount) by day, rating, and city. Remember to join with the appropriate dimension tables to replace the keys with the dimension labels. Sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below.

In [8]:
%%time
%%sql

SELECT day, rating, city, SUM(sales_amount) AS revenue
FROM factsales
JOIN dimdate ON (factsales.date_key=dimdate.date_key)
JOIN dimmovie ON (factsales.movie_key=dimmovie.movie_key)
JOIN dimCustomer ON (dimCustomer.customer_key = factSales.customer_key)
GROUP BY day, rating, city
ORDER BY revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.23 ms, sys: 677 µs, total: 4.91 ms
Wall time: 36.6 ms


[(30, 'G', 'San Bernardino', Decimal('24.97')),
 (30, 'NC-17', 'Apeldoorn', Decimal('23.95')),
 (21, 'NC-17', 'Belm', Decimal('22.97')),
 (28, 'R', 'Mwanza', Decimal('21.97')),
 (21, 'G', 'Citt del Vaticano', Decimal('21.97')),
 (30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (1, 'R', 'Qomsheh', Decimal('19.97')),
 (22, 'R', 'Yangor', Decimal('19.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (17, 'G', 'Rajkot', Decimal('19.97')),
 (19, 'PG', 'Najafabad', Decimal('19.96')),
 (30, 'R', 'Fengshan', Decimal('19.95')),
 (28, 'PG', 'So Leopoldo', Decimal('18.98')),
 (21, 'G', 'Wroclaw', Decimal('18.98')),
 (1, 'NC-17', 'Memphis', Decimal('18.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (19, 'PG', 'Sokoto', Decimal('18.97')),
 (30, 'G', 'Omdurman', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

TODO: Write a query that reduces the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows. The first few rows of your output should match the table below. 

In [12]:
%%time
%%sql

SELECT dimdate.day, dimmovie.rating, dimcustomer.city, SUM(factsales.sales_amount) AS revenue
FROM factsales
JOIN dimdate ON (factsales.date_key=dimdate.date_key)
JOIN dimmovie ON (factsales.movie_key=dimmovie.movie_key)
JOIN dimcustomer ON (factsales.customer_key=dimcustomer.customer_key)
WHERE dimmovie.rating='PG-13'
GROUP BY dimdate.day, dimmovie.rating, dimcustomer.city
ORDER BY revenue DESC

 * postgresql://student:***@127.0.0.1:5432/pagila_star
3107 rows affected.
CPU times: user 10.4 ms, sys: 0 ns, total: 10.4 ms
Wall time: 22.5 ms


[(30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95')),
 (20, 'PG-13', 'Baha Blanca', Decimal('17.98')),
 (30, 'PG-13', 'Nagareyama', Decimal('17.98')),
 (21, 'PG-13', 'Parbhani', Decimal('17.98')),
 (30, 'PG-13', 'Tanauan', Decimal('17.96')),
 (17, 'PG-13', 'Ikerre', Decimal('17.95')),
 (30, 'PG-13', 'Zhoushan', Decimal('16.98')),
 (30, 'PG-13', 'Yerevan', Decimal('16.97')),
 (30, 'PG-13', 'Newcastle', Decimal('16.97')),
 (20, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (30, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (17, 'PG-13', 'Yantai', Decimal('15.98')),
 (8, 'PG-13', 'Pontianak', Decimal('15.98')),
 (20, 'PG-13', 'Johannesburg', Decimal('15.97')),
 (30, 'PG-13', 'Toulouse', Decimal('15.97')),
 (23, 'PG-13', 'Tiefa', Decimal('15.97')),
 (10, 'PG-13', 'Udaipur', Decimal('15.97')),
 (1, 'PG-13', 'Moscow', Decimal('15.97')),
 (19, 'PG-13', 'Owo', Decimal('14.98')),
 (29, 'PG-13', 'Gaziantep', Decimal('14.98')),
 (19, 'PG-13', 'Touliu', Decimal('14.98')),
 (21, 'PG-13', 'Aden', Decimal('14.98')),
 (27, 'PG-13', 'Sawhaj', Decimal('14.98')),
 (18, 'PG-13', 'Okayama', Decimal('14.98')),
 (27, 'PG-13', 'Gulbarga', Decimal('14.98')),
 (12, 'PG-13', 'Usolje-Sibirskoje', Decimal('14.98')),
 (27, 'PG-13', 'Phnom Penh', Decimal('14.97')),
 (18, 'PG-13', 'Cavite', Decimal('14.97')),
 (20, 'PG-13', 'Ciparay', Decimal('14.97')),
 (18, 'PG-13', 'Shimonoseki', Decimal('14.97')),
 (29, 'PG-13', 'Jedda', Decimal('13.98')),
 (29, 'PG-13', 'Maikop', Decimal('13.98')),
 (1, 'PG-13', 'Haiphong', Decimal('13.98')),
 (27, 'PG-13', 'Idfu', Decimal('13.98')),
 (9, 'PG-13', 'Baicheng', Decimal('13.98')),
 (29, 'PG-13', 'Cuauhtmoc', Decimal('13.98')),
 (21, 'PG-13', 'Apeldoorn', Decimal('13.98')),
 (17, 'PG-13', 'Tanza', Decimal('13.98')),
 (11, 'PG-13', 'Boa Vista', Decimal('13.98')),
 (21, 'PG-13', 'Lublin', Decimal('13.98')),
 (15, 'PG-13', 'Valparai', Decimal('13.98')),
 (29, 'PG-13', 'Qomsheh', Decimal('13.98')),
 (20, 'PG-13', 'Naju', Decimal('13.98')),
 (30, 'PG-13', 'Boksburg', Decimal('13.97')),
 (21, 'PG-13', 'Maracabo', Decimal('13.97')),
 (7, 'PG-13', 'Avellaneda', Decimal('13.97')),
 (20, 'PG-13', 'Boa Vista', Decimal('13.96')),
 (9, 'PG-13', 'Mysore', Decimal('12.98')),
 (18, 'PG-13', 'Ibirit', Decimal('12.98')),
 (18, 'PG-13', 'Yangor', Decimal('12.98')),
 (7, 'PG-13', 'La Paz', Decimal('12.98')),
 (21, 'PG-13', 'Firozabad', Decimal('12.98')),
 (11, 'PG-13', 'Ife', Decimal('12.98')),
 (19, 'PG-13', 'Pemalang', Decimal('12.98')),
 (17, 'PG-13', 'Usolje-Sibirskoje', Decimal('12.98')),
 (18, 'PG-13', 'Kimberley', Decimal('12.98')),
 (21, 'PG-13', 'Etawah', Decimal('12.98')),
 (5, 'PG-13', 'Nha Trang', Decimal('12.98')),
 (18, 'PG-13', 'Udaipur', Decimal('12.98')),
 (17, 'PG-13', 's-Hertogenbosch', Decimal('12.98')),
 (29, 'PG-13', 'Aden', Decimal('12.98')),
 (21, 'PG-13', 'Masqat', Decimal('12.98')),
 (27, 'PG-13', 'Quilmes', Decimal('12.97')),
 (8, 'PG-13', 'Bucuresti', Decimal('12.97')),
 (30, 'PG-13', 'Shaoguan', Decimal('12.97')),
 (22, 'PG-13', 'Ondo', Decimal('12.97')),
 (26, 'PG-13', 'Iwaki', Decimal('12.97')),
 (20, 'PG-13', 'Vilnius', Decimal('12.97')),
 (25, 'PG-13', 'Mannheim', Decimal('11.99')),
 (21, 'PG-13', 'Pemalang', Decimal('11.99')),
 (7, 'PG-13', 'Juiz de Fora', Decimal('11.99')),
 (30, 'PG-13', 'Lhokseumawe', Decimal('11.98')),
 (10, 'PG-13', 'San Felipe del Progreso', Decimal('11.98')),
 (20, 'PG-13', 'Maring', Decimal('11.98')),
 (30, 'PG-13', 'Apeldoorn', Decimal('11.98')),
 (18, 'PG-13', 'Uluberia', Decimal('11.98')),
 (1, 'PG-13', 'Iwakuni', Decimal('11.98')),
 (2, 'PG-13', 'Antofagasta', Decimal('11.98')),
 (21, 'PG-13', 'Tychy', Decimal('11.98')),
 (29, 'PG-13', 'Southport', Decimal('11.98')),
 (30, 'PG-13', 'La Romana', Decimal('11.98')),
 (9, 'PG-13', 'Kaliningrad', Decimal('11.98')),
 (27, 'PG-13', 'Vijayawada', Decimal('11.98')),
 (21, 'PG-13

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

TODO: Write a query to create a subcube of the initial cube that includes moves with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

The first few rows of your output should match the table below. 

In [14]:
%%time
%%sql

SELECT dimdate.day, dimmovie.rating, dimcustomer.city, SUM(factsales.sales_amount) AS revenue
FROM factsales
JOIN dimdate ON (factsales.date_key=dimdate.date_key)
JOIN dimmovie ON (factsales.movie_key=dimmovie.movie_key)
JOIN dimcustomer ON (factsales.customer_key=dimcustomer.customer_key)
WHERE dimmovie.rating IN('PG-13','PG') 
AND dimcustomer.city IN ('Bellevue','Lancaster')
AND dimdate.day IN (1, 15, 30)
GROUP BY dimdate.day, dimmovie.rating, dimcustomer.city
ORDER BY revenue DESC

 * postgresql://student:***@127.0.0.1:5432/pagila_star
6 rows affected.
CPU times: user 2.35 ms, sys: 3.31 ms, total: 5.66 ms
Wall time: 10.1 ms


[(30, 'PG', 'Lancaster', Decimal('12.98')),
 (1, 'PG-13', 'Lancaster', Decimal('5.99')),
 (30, 'PG-13', 'Bellevue', Decimal('3.99')),
 (30, 'PG-13', 'Lancaster', Decimal('2.99')),
 (15, 'PG-13', 'Bellevue', Decimal('1.98')),
 (1, 'PG', 'Bellevue', Decimal('0.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>